File: data structures and algorithms/cpp/08_052.cpp
#include <string>
#include <vector>
#include <set>
#include <stack>
#include <unordered_map>

class Solution {
public:
    std::vector<std::string> generateBracketExpressions(
            const std::string& expr) {
        std::set<std::string> result;
        std::vector<std::string> brackets = {"()", "[]", "{}"};
        
        // Check if we can insert a bracket pair at each position
        for (int i = 0; i <= expr.length(); i++) {
            for (const auto& bracket : brackets) {
                std::string newExpr = expr.substr(0, i) + 
                                     bracket[0] + 
                                     expr.substr(i);
                
                // Try all positions for the closing bracket
                for (int j = i + 1; j <= newExpr.length(); j++) {
                    std::string candidate = newExpr.substr(0, j) + 
                                           bracket[1] + 
                                           newExpr.substr(j);
                    if (isValid(candidate)) {
                        result.insert(candidate);
                    }
                }
            }
        }
        
        // Add each type of bracket pair around the entire expr
        for (const auto& bracket : brackets) {
            result.insert(bracket[0] + expr + bracket[1]);
        }
        
        return std::vector<std::string>(
                result.begin(), result.end());
    }
    
private:
    bool isValid(const std::string& expr) {
        std::stack<char> stack;
        std::unordered_map<char, char> bracketMap = {
            {')', '('},
            {']', '['},
            {'}', '{'}
        };
        
        for (char c : expr) {
            if (c == '(' || c == '[' || c == '{') {
                stack.push(c);
            } else if (c == ')' || c == ']' || c == '}') {
                if (stack.empty() || 
                    stack.top() != bracketMap[c]) {
                    return false;
                }
                stack.pop();
            }
        }
        
        return stack.empty();
    }
};

File: data structures and algorithms/cs/08_052.cs
using System;
using System.Collections.Generic;

public class Solution
{
   public List<string> GenerateBracketExpressions(
      string expr)
   {
      var result = new HashSet<string>();
      string[] brackets = { "()", "[]", "{}" };

      for (int i = 0; i <= expr.Length; i++)
      {
         foreach (var bracket in brackets)
         {
            string newExpr = expr.Substring(0, i)
               + bracket[0] + expr.Substring(i);

            for (int j = i + 1; j <= newExpr.Length; j++)
            {
               string candidate = newExpr.Substring(0, j)
                  + bracket[1] + newExpr.Substring(j);
               if (IsValid(candidate))
               {
                  result.Add(candidate);
               }
            }
         }
      }

      foreach (var bracket in brackets)
      {
         string wrapped = bracket[0] + expr + bracket[1];
         if (IsValid(wrapped))
         {
            result.Add(wrapped);
         }
      }

      return new List<string>(result);
   }

   private bool IsValid(string expr)
   {
      var stack = new Stack<char>();
      var bracketMap = new Dictionary<char, char> {
         { ')', '(' },
         { ']', '[' },
         { '}', '{' }
      };

      foreach (var c in expr)
      {
         if (c == '(' || c == '[' || c == '{')
         {
            stack.Push(c);
         }
         else if (c == ')' || c == ']' || c == '}')
         {
            if (stack.Count == 0 || stack.Pop()
               != bracketMap[c])
            {
               return false;
            }
         }
      }

      return stack.Count == 0;
   }
}